packages/flutter/lib/src/cupertino/refresh.dart

```diff
@@ -0,0 +1,514 @@
+import 'package:flutter/rendering.dart';
+import 'package:flutter/scheduler.dart';
+import 'package:flutter/services.dart';
+import 'package:flutter/widgets.dart';
+import 'activity_indicator.dart';
+import 'colors.dart';
+import 'icons.dart';
+
+class _CupertinoRefreshSliver extends SingleChildRenderObjectWidget {
+  const _CupertinoRefreshSliver({
+    this.refreshIndicatorLayoutExtent: 0.0,
+    this.hasLayoutExtent,
+    Widget child,
+  }) : super(child: child);
+
+  /// The amount of space the indicator should occupy in the sliver in a
+  /// resting state when in the refreshing mode.
+  final double refreshIndicatorLayoutExtent;
+  /// _RenderCupertinoRefreshSliver will paint the child in the available
+  /// space either way but this instructs the _RenderCupertinoRefreshSliver
+  /// on whether to also occupy any layoutExtent space or not.
+  final bool hasLayoutExtent;
+
+  @override
+  _RenderCupertinoRefreshSliver createRenderObject(BuildContext context) {
+    return new _RenderCupertinoRefreshSliver(
+      refreshIndicatorExtent: refreshIndicatorLayoutExtent,
+      hasLayoutExtent: hasLayoutExtent,
+    );
+  }
+
+  @override
+  void updateRenderObject(BuildContext context, covariant _RenderCupertinoRefreshSliver renderObject) {
+    renderObject
+        ..refreshIndicatorLayoutExtent = refreshIndicatorLayoutExtent
+        ..hasLayoutExtent = hasLayoutExtent;
+  }
+}
+
+/// RenderSliver object that gives its child RenderBox object space to paint
+/// in the overscrolled gap and may or may not hold that overscrolled gap
+/// around the RenderBox depending on whether [layoutExtent] is set.
+///
+/// The [layoutExtentOffsetCompensation] field keeps internal accounting to
+/// prevent scroll position jumps as the [layoutExtent] is set and unset.
+class _RenderCupertinoRefreshSliver
+    extends RenderSliver
+    with RenderObjectWithChildMixin<RenderBox> {
+  _RenderCupertinoRefreshSliver({
+    double refreshIndicatorExtent,
+    bool hasLayoutExtent,
+    RenderBox child,
+  }) : _refreshIndicatorExtent = refreshIndicatorExtent,
+       _hasLayoutExtent = hasLayoutExtent {
+    this.child = child;
+  }
+
+  /// The amount of layout space the indicator should occupy in the sliver in a
+  /// resting state when in the refreshing mode.
+  double get refreshIndicatorLayoutExtent => _refreshIndicatorExtent;
+  double _refreshIndicatorExtent;
+  set refreshIndicatorLayoutExtent(double value) {
+    if (value == _refreshIndicatorExtent)
+      return;
+    _refreshIndicatorExtent = value;
+    markNeedsLayout();
+  }
+
+  /// The child box will be laid out and painted in the available space either
+  /// way but this determines whether to also occupy any layoutExtent space or
+  /// not.
+  bool get hasLayoutExtent => _hasLayoutExtent;
+  bool _hasLayoutExtent;
+  set hasLayoutExtent(bool value) {
+    if (value == _hasLayoutExtent)
+      return;
+    _hasLayoutExtent = value;
+    markNeedsLayout();
+  }
+
```

Comment:
if these properties can't be null, assert non-null in the constructor and the setters

Reply:
Done